RealmSwift/SyncSubscription.swift

```diff
@@ -0,0 +1,250 @@
+////////////////////////////////////////////////////////////////////////////
+ //
+ // Copyright 2021 Realm Inc.
+ //
+ // Licensed under the Apache License, Version 2.0 (the "License");
+ // you may not use this file except in compliance with the License.
+ // You may obtain a copy of the License at
+ //
+ // http://www.apache.org/licenses/LICENSE-2.0
+ //
+ // Unless required by applicable law or agreed to in writing, software
+ // distributed under the License is distributed on an "AS IS" BASIS,
+ // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ // See the License for the specific language governing permissions and
+ // limitations under the License.
+ //
+ ////////////////////////////////////////////////////////////////////////////
+
+import Foundation
+
+// State Updates
+// Some operations will return a `SubscriptionTask` which can be used to get state updates (There will be a Combine API as well not described here)
+internal enum SyncSubscriptionState {
+    // Subscription is complete and the server is in "steady-state" synchronization.
+    case complete
+    // The Subscription encountered an error.
+    case error(Error)
+    // The server is processing the subscription and updating the Realm data
+    // with new matches
+    case pending
+}
+
+extension SyncSubscriptionState: Equatable {
+    static func == (lhs: SyncSubscriptionState, rhs: SyncSubscriptionState) -> Bool {
+        true
+    }
+}
+
+internal protocol AnySyncSubscription {}
+
+internal class SyncSubscription<Element: Object>: AnySyncSubscription {
+    internal typealias QueryFunction = (Query<Element>) -> Query<Element>
+
+    // When the subscription was created. Recorded automatically.
+    internal var createdAt: Date = Date()
+
+    // When the subscription was last updated. Recorded automatically.
+    internal var updatedAt: Date = Date()
+
+    // Name of the subscription, if not specified it will return the value in Query
+    internal var name: String = ""
+
+    #if swift(>=5.4)
+    // Update query for subscription
+    internal func update(@QueryBuilder _ to: () -> (AnySyncSubscription)) throws {
```

Comment:
We are missing a way to update queries for swift <5.4.

Reply:
I'm adding this for now because we are supporting Xcode 12.2 but referencing this PR https://github.com/realm/realm-cocoa/pull/7501 this is something that is going to change in the near future, do we want to add another API?